daemon/containerd/image_builder.go

```diff
@@ -307,22 +372,33 @@ func (rw *rwlayer) Commit() (builder.ROLayer, error) {
 		return nil, err
 	}
 
+	if err := lm.AddResource(ctx, lease, leases.Resource{
+		ID:   desc.Digest.String(),
+		Type: "content",
+	}); err != nil {
+		return nil, fmt.Errorf("failed to add diff content to lease")
+	}
+
 	return &rolayer{
 		key:                key,
 		c:                  rw.c,
 		snapshotter:        rw.snapshotter,
 		diffID:             diffID,
 		contentStoreDigest: desc.Digest,
+		lease:              &lease,
 	}, nil
 }
 
 func (rw *rwlayer) Release() error {
-	snapshotter := rw.c.SnapshotService(rw.snapshotter)
-	err := snapshotter.Remove(context.TODO(), rw.key)
-	if err != nil && !cerrdefs.IsNotFound(err) {
-		return err
+	if rw.lease != nil {
+		lm := rw.c.LeasesService()
+		err := lm.Delete(context.TODO(), *rw.lease)
+		if err != nil {
+			log.G(context.TODO()).WithError(err).Warn("failed to delete lease when releasing RWLayer")
+		} else {
+			rw.lease = nil
+		}
 	}
```

Comment:
sorry, one more question; I see here we're releasing the lease, _after_ which we try to do the unmounting and removing (which may fail).

- Are there any consequences of releasing the lease _before_ we perform those actions?
- Should releasing be done afterwards? (which could be in a defer)
- And (not sure) should it only be released on "success" or also on "error" ?

Reply:
Ah, good catch. Yes, the lease should be performed later.